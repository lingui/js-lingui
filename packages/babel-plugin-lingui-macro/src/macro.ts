import { createMacro, MacroParams } from "babel-plugin-macros"

import type { VisitNodeObject } from "@babel/traverse"
import { Program } from "@babel/types"

import linguiPlugin from "./index"
import { JsMacroName, JsxMacroName } from "./constants"

function macro({ state, babel, config }: MacroParams) {
  if (!state.get("linguiProcessed")) {
    state.opts = config
    const plugin = linguiPlugin(babel)

    const { enter, exit } = plugin.visitor.Program as VisitNodeObject<
      any,
      Program
    >

    enter(state.file.path, state)
    state.file.path.traverse(plugin.visitor, state)
    exit(state.file.path, state)

    state.set("linguiProcessed", true)
  }

  return { keepImports: true }
}

;[
  JsMacroName.defineMessage,
  JsMacroName.msg,
  JsMacroName.t,
  JsMacroName.useLingui,
  JsMacroName.plural,
  JsMacroName.select,
  JsMacroName.selectOrdinal,

  JsxMacroName.Trans,
  JsxMacroName.Plural,
  JsxMacroName.Select,
  JsxMacroName.SelectOrdinal,
].forEach((name) => {
  Object.defineProperty(module.exports, name, {
    get() {
      throw new Error(
        `The macro you imported from "@lingui/core/macro" or "@lingui/react/macro" is being executed outside the context of compilation with babel-plugin-macros. ` +
          `This indicates that you don't have the babel plugin "babel-plugin-macros" configured correctly. ` +
          `Please see the documentation for how to configure babel-plugin-macros properly: ` +
          "https://github.com/kentcdodds/babel-plugin-macros/blob/main/other/docs/user.md"
      )
    },
  })
})

export default createMacro(macro, {
  configName: "lingui",
}) as { isBabelMacro: true }
