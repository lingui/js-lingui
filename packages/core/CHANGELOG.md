# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [4.8.0-next.0](https://github.com/lingui/js-lingui/compare/v4.7.1...v4.8.0-next.0) (2024-03-04)

**Note:** Version bump only for package @lingui/core

## [4.7.1](https://github.com/lingui/js-lingui/compare/v4.7.0...v4.7.1) (2024-02-20)

### Bug Fixes

- **core:** use named instead of default export for unraw lib ([#1837](https://github.com/lingui/js-lingui/issues/1837)) ([85f0944](https://github.com/lingui/js-lingui/commit/85f094449ea1b8aa46339b2975b7e02fff4df234))
- don't replace octothorpe coming from variable ([#1850](https://github.com/lingui/js-lingui/issues/1850)) ([e321729](https://github.com/lingui/js-lingui/commit/e321729702b117c8cfea5ae861cf3767030cc9c4))

# [4.7.0](https://github.com/lingui/js-lingui/compare/v4.6.0...v4.7.0) (2024-01-05)

**Note:** Version bump only for package @lingui/core

# [4.6.0](https://github.com/lingui/js-lingui/compare/v4.5.0...v4.6.0) (2023-12-01)

### Features

- **message-utils:** make generateMessageId to be working in browser ([#1776](https://github.com/lingui/js-lingui/issues/1776)) ([f879ddb](https://github.com/lingui/js-lingui/commit/f879ddbbc4627f94c579e0156958b4ec4026e371))

# [4.5.0](https://github.com/lingui/js-lingui/compare/v4.4.2...v4.5.0) (2023-09-14)

**Note:** Version bump only for package @lingui/core

## [4.4.2](https://github.com/lingui/js-lingui/compare/v4.4.1...v4.4.2) (2023-08-31)

**Note:** Version bump only for package @lingui/core

## [4.4.1](https://github.com/lingui/js-lingui/compare/v4.4.0...v4.4.1) (2023-08-30)

**Note:** Version bump only for package @lingui/core

# [4.4.0](https://github.com/lingui/js-lingui/compare/v4.3.0...v4.4.0) (2023-08-08)

**Note:** Version bump only for package @lingui/core

# [4.3.0](https://github.com/lingui/js-lingui/compare/v4.2.1...v4.3.0) (2023-06-29)

**Note:** Version bump only for package @lingui/core

## [4.2.1](https://github.com/lingui/js-lingui/compare/v4.2.0...v4.2.1) (2023-06-07)

### Bug Fixes

- **core:** don't crash on escaped unicode symbols in messages ([#1692](https://github.com/lingui/js-lingui/issues/1692)) ([461c2fc](https://github.com/lingui/js-lingui/commit/461c2fc533647a877ea3a9b14e81544049e4e2b2))

# [4.2.0](https://github.com/lingui/js-lingui/compare/v4.1.2...v4.2.0) (2023-05-26)

### Bug Fixes

- fallback to `other` only when undefined ([#1665](https://github.com/lingui/js-lingui/issues/1665)) ([a79de75](https://github.com/lingui/js-lingui/commit/a79de757df14249f76945da9fd9ce529129f2269))

## [4.1.2](https://github.com/lingui/js-lingui/compare/v4.1.1...v4.1.2) (2023-05-17)

**Note:** Version bump only for package @lingui/core

## [4.1.1](https://github.com/lingui/js-lingui/compare/v4.1.0...v4.1.1) (2023-05-17)

**Note:** Version bump only for package @lingui/core

# [4.1.0](https://github.com/lingui/js-lingui/compare/v4.0.0...v4.1.0) (2023-05-15)

### Bug Fixes

- format should follow locale order ([#1619](https://github.com/lingui/js-lingui/issues/1619)) ([4d53b7b](https://github.com/lingui/js-lingui/commit/4d53b7b64f11968dca8332655a03237d70c6eb6e))

## [3.17.2](https://github.com/lingui/js-lingui/compare/v3.17.1...v3.17.2) (2023-02-24)

### Bug Fixes

- param name for `missing` in `setupI18nProps` ([#1411](https://github.com/lingui/js-lingui/issues/1411)) ([6459f02](https://github.com/lingui/js-lingui/commit/6459f024fa74370f95a77738b7d72f114a53c9db))

## [3.17.1](https://github.com/lingui/js-lingui/compare/v3.17.0...v3.17.1) (2023-02-07)

**Note:** Version bump only for package @lingui/core

# [3.17.0](https://github.com/lingui/js-lingui/compare/v3.16.1...v3.17.0) (2023-02-01)

**Note:** Version bump only for package @lingui/core

## [3.16.1](https://github.com/lingui/js-lingui/compare/v3.16.0...v3.16.1) (2023-01-24)

### Bug Fixes

- **core:** sub entry export for react-native ([#1350](https://github.com/lingui/js-lingui/issues/1350)) ([c178ce3](https://github.com/lingui/js-lingui/commit/c178ce3cc91e64d190965dc684df74162cc17d02))

# [3.16.0](https://github.com/lingui/js-lingui/compare/v3.15.0...v3.16.0) (2023-01-18)

**Note:** Version bump only for package @lingui/core

# [3.15.0](https://github.com/lingui/js-lingui/compare/v3.14.0...v3.15.0) (2022-11-07)

### Bug Fixes

- Pass formatter to pluralization function ([#1294](https://github.com/lingui/js-lingui/issues/1294)) ([33af3ba](https://github.com/lingui/js-lingui/commit/33af3ba3dc22bd4128d98e061cb9d97f75c6503a))

# [3.14.0](https://github.com/lingui/js-lingui/compare/v3.13.3...v3.14.0) (2022-06-22)

### Bug Fixes

- Add extra package.json under `/esm` ([#1258](https://github.com/lingui/js-lingui/issues/1258)) ([80cd337](https://github.com/lingui/js-lingui/commit/80cd3378ceb5677bfa50b14f67e4e31703392298))

## [3.13.3](https://github.com/lingui/js-lingui/compare/v3.13.2...v3.13.3) (2022-04-24)

**Note:** Version bump only for package @lingui/core

## [3.13.2](https://github.com/lingui/js-lingui/compare/v3.13.1...v3.13.2) (2022-01-24)

### Bug Fixes

- localeData interpolation broken ([457735b](https://github.com/lingui/js-lingui/commit/457735bb61f0a75d0ce176874b1c67a65c4b3084))

## [3.13.1](https://github.com/lingui/js-lingui/compare/v3.13.0...v3.13.1) (2022-01-21)

### Bug Fixes

- i18n.activate should load instantly new messages ([#1182](https://github.com/lingui/js-lingui/issues/1182)) ([f8f47a2](https://github.com/lingui/js-lingui/commit/f8f47a2385fe3d8dd3395c493027de2492509325))

# [3.13.0](https://github.com/lingui/js-lingui/compare/v3.12.1...v3.13.0) (2021-11-26)

### Features

- msgctxt support ([#1094](https://github.com/lingui/js-lingui/issues/1094)) ([8ee42cb](https://github.com/lingui/js-lingui/commit/8ee42cbfe26bc6d055748dcf2713ab8ade7ec827))

## [3.12.1](https://github.com/lingui/js-lingui/compare/v3.12.0...v3.12.1) (2021-09-28)

**Note:** Version bump only for package @lingui/core

# [3.12.0](https://github.com/lingui/js-lingui/compare/v3.11.1...v3.12.0) (2021-09-28)

**Note:** Version bump only for package @lingui/core

## [3.11.1](https://github.com/lingui/js-lingui/compare/v3.11.0...v3.11.1) (2021-09-07)

**Note:** Version bump only for package @lingui/core

# [3.11.0](https://github.com/lingui/js-lingui/compare/v3.10.4...v3.11.0) (2021-09-07)

### Bug Fixes

- sideEffects to false for shrinking bundle size ([#1129](https://github.com/lingui/js-lingui/issues/1129)) ([57cd2e5](https://github.com/lingui/js-lingui/commit/57cd2e576945ba30aea30d5cf5bcb27d1f77fe4c))
- unicode chars were removed from .po files when extracting ([#1125](https://github.com/lingui/js-lingui/issues/1125)) ([d688329](https://github.com/lingui/js-lingui/commit/d688329c0457a080d5cd485a95e94ffa5e00c5ae))

## [3.10.4](https://github.com/lingui/js-lingui/compare/v3.10.3...v3.10.4) (2021-06-16)

**Note:** Version bump only for package @lingui/core

## [3.10.3](https://github.com/lingui/js-lingui/compare/v3.10.2...v3.10.3) (2021-06-14)

### Bug Fixes

- add string return type to i18n.\_ ([#1081](https://github.com/lingui/js-lingui/issues/1081)) ([68d32d6](https://github.com/lingui/js-lingui/commit/68d32d63899c9a0b2c66891070831c6488a7d013))

## [3.10.2](https://github.com/lingui/js-lingui/compare/v3.10.1...v3.10.2) (2021-06-08)

**Note:** Version bump only for package @lingui/core

## [3.10.1](https://github.com/lingui/js-lingui/compare/v3.10.0...v3.10.1) (2021-06-08)

### Bug Fixes

- reverts reduce size of ESM packages ([#1066](https://github.com/lingui/js-lingui/issues/1066)) ([3a057e0](https://github.com/lingui/js-lingui/commit/3a057e0c61224b98c93203e0d88136fa48f309ba))

# [3.10.0](https://github.com/lingui/js-lingui/compare/v3.9.0...v3.10.0) (2021-06-08)

### Bug Fixes

- reduce size of ESM packages ([#1066](https://github.com/lingui/js-lingui/issues/1066)) ([9990eba](https://github.com/lingui/js-lingui/commit/9990ebaa9d30f7e218c106a2abfd7ddbcf0e0170))

# [3.9.0](https://github.com/lingui/js-lingui/compare/v3.8.10...v3.9.0) (2021-05-18)

### Bug Fixes

- ESM dev entrypoint was causing unexpected behaviors ([#1059](https://github.com/lingui/js-lingui/issues/1059)) ([2b8d70b](https://github.com/lingui/js-lingui/commit/2b8d70b35ed103d94a2f88ea7c6813a29a1d5635))

## [3.8.10](https://github.com/lingui/js-lingui/compare/v3.8.9...v3.8.10) (2021-04-19)

**Note:** Version bump only for package @lingui/core

## [3.8.9](https://github.com/lingui/js-lingui/compare/v3.8.8...v3.8.9) (2021-04-09)

### Bug Fixes

- event emitter refactor (reverted) ([#1038](https://github.com/lingui/js-lingui/issues/1038)) ([f299493](https://github.com/lingui/js-lingui/commit/f299493999299fe9a7d0e01b9045e7f0a9813c6a))

## [3.8.8](https://github.com/lingui/js-lingui/compare/v3.8.7...v3.8.8) (2021-04-09)

**Note:** Version bump only for package @lingui/core

## [3.8.7](https://github.com/lingui/js-lingui/compare/v3.8.6...v3.8.7) (2021-04-09)

### Bug Fixes

- unicode chars in native environments + event emitter refactor ([#1036](https://github.com/lingui/js-lingui/issues/1036)) ([39fa90d](https://github.com/lingui/js-lingui/commit/39fa90d95c08f105f3f7feb17b65d9b8f916b73a))

## [3.8.6](https://github.com/lingui/js-lingui/compare/v3.8.5...v3.8.6) (2021-04-08)

**Note:** Version bump only for package @lingui/core

## [3.8.5](https://github.com/lingui/js-lingui/compare/v3.8.4...v3.8.5) (2021-04-08)

**Note:** Version bump only for package @lingui/core

## [3.8.4](https://github.com/lingui/js-lingui/compare/v3.8.3...v3.8.4) (2021-04-08)

### Bug Fixes

- undefined interpolation on t macro ([#1030](https://github.com/lingui/js-lingui/issues/1030)) ([194f9b3](https://github.com/lingui/js-lingui/commit/194f9b318434bcaf4b68af549a6ed246fed2be02))

## [3.8.3](https://github.com/lingui/js-lingui/compare/v3.8.2...v3.8.3) (2021-04-05)

**Note:** Version bump only for package @lingui/core

## [3.8.2](https://github.com/lingui/js-lingui/compare/v3.8.1...v3.8.2) (2021-03-31)

**Note:** Version bump only for package @lingui/core

## [3.8.1](https://github.com/lingui/js-lingui/compare/v3.8.0...v3.8.1) (2021-03-23)

**Note:** Version bump only for package @lingui/core

# [3.8.0](https://github.com/lingui/js-lingui/compare/v3.7.2...v3.8.0) (2021-03-23)

**Note:** Version bump only for package @lingui/core

## [3.7.2](https://github.com/lingui/js-lingui/compare/v3.7.1...v3.7.2) (2021-03-14)

**Note:** Version bump only for package @lingui/core

## [3.7.1](https://github.com/lingui/js-lingui/compare/v3.7.0...v3.7.1) (2021-03-07)

### Bug Fixes

- exports attribute on package.json, expo compat ([#997](https://github.com/lingui/js-lingui/issues/997)) ([79aa509](https://github.com/lingui/js-lingui/commit/79aa50998185847064d80f2a38be0ebcb64424f9))

# [3.7.0](https://github.com/lingui/js-lingui/compare/v3.6.0...v3.7.0) (2021-03-04)

### Features

- raise event on missing translation ([#993](https://github.com/lingui/js-lingui/issues/993)) ([ecf83c3](https://github.com/lingui/js-lingui/commit/ecf83c37e177cb10f157fb8913d73e18662e3ca4))

# [3.6.0](https://github.com/lingui/js-lingui/compare/v3.5.1...v3.6.0) (2021-02-23)

### Features

- ship universal modules with ESM ([#979](https://github.com/lingui/js-lingui/issues/979)) ([6cd5fe0](https://github.com/lingui/js-lingui/commit/6cd5fe0a71dd5cf7e0832bd3e9902a2f6ba789f6))

## [3.5.1](https://github.com/lingui/js-lingui/compare/v3.5.0...v3.5.1) (2021-02-09)

**Note:** Version bump only for package @lingui/core

# [3.5.0](https://github.com/lingui/js-lingui/compare/v3.4.0...v3.5.0) (2021-02-02)

**Note:** Version bump only for package @lingui/core

# [3.4.0](https://github.com/lingui/js-lingui/compare/v3.3.0...v3.4.0) (2021-01-13)

### Bug Fixes

- added return type to i18n.\_ ([#922](https://github.com/lingui/js-lingui/issues/922)) ([249c486](https://github.com/lingui/js-lingui/commit/249c486d258a4fb44bae1e3da1765e00003429a7))

# [3.3.0](https://github.com/lingui/js-lingui/compare/v3.2.3...v3.3.0) (2020-12-08)

**Note:** Version bump only for package @lingui/core

## [3.2.3](https://github.com/lingui/js-lingui/compare/v3.2.2...v3.2.3) (2020-11-22)

**Note:** Version bump only for package @lingui/core

## [3.2.2](https://github.com/lingui/js-lingui/compare/v3.2.1...v3.2.2) (2020-11-20)

**Note:** Version bump only for package @lingui/core

# [3.2.0](https://github.com/lingui/js-lingui/compare/v3.1.0...v3.2.0) (2020-11-12)

### Bug Fixes

- MessageDescriptor values type ([#848](https://github.com/lingui/js-lingui/issues/848)) ([9712d94](https://github.com/lingui/js-lingui/commit/9712d94d043c8d40cbc5d017474b5938eb02f8d6))

# [3.1.0](https://github.com/lingui/js-lingui/compare/v3.0.3...v3.1.0) (2020-11-10)

### Bug Fixes

- improved performance of formatters ([#818](https://github.com/lingui/js-lingui/issues/818)) ([22667ad](https://github.com/lingui/js-lingui/commit/22667adba5b07cc94abacff8e8b5f5b19202576c))
- show error when plurals aren't loaded ([#824](https://github.com/lingui/js-lingui/issues/824)) ([296b6a1](https://github.com/lingui/js-lingui/commit/296b6a1a1f332064f040cc987c4359411d307258))

### Features

- use fallback locales from cldr ([#820](https://github.com/lingui/js-lingui/issues/820)) ([2d9e124](https://github.com/lingui/js-lingui/commit/2d9e124b91f1ba7a65e9f997a3ba952679c6c23a))
