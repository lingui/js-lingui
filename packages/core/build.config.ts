import { defineBuildConfig } from "unbuild"

export default defineBuildConfig({
  externals: ["@lingui/core"],
  failOnWarn: false,
})
