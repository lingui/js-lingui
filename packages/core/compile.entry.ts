/**
 * this is Rollup Entry
 */
export * from "./src/compile"
