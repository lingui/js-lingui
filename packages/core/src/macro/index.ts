import { createMacro, MacroParams } from "babel-plugin-macros"
import { getConfig as loadConfig, LinguiConfigNormalized } from "@lingui/conf"

import { MacroJs } from "./macroJs"
import type { NodePath } from "@babel/traverse"
import {
  addImport,
  isRootPath,
  throwNonMacroContextError,
} from "@lingui/macro-lib"

export type LinguiMacroOpts = {
  // explicitly set by CLI when running extraction process
  extract?: boolean
  linguiConfig?: LinguiConfigNormalized
}

export const jsMacroTags = new Set([
  "defineMessage",
  "arg",
  "t",
  "plural",
  "select",
  "selectOrdinal",
])

let config: LinguiConfigNormalized

function getConfig(_config?: LinguiConfigNormalized) {
  if (_config) {
    config = _config
  }
  if (!config) {
    config = loadConfig()
  }
  return config
}

export function macro({ references, state, babel, config }: MacroParams) {
  const opts: LinguiMacroOpts = config as LinguiMacroOpts

  const { i18nImportModule, i18nImportName } = getConfig(
    opts.linguiConfig
  ).runtimeConfigModule

  const jsNodes = new Set<NodePath>()
  let needsI18nImport = false

  Object.keys(references).forEach((tagName) => {
    const nodes = references[tagName]

    if (jsMacroTags.has(tagName)) {
      nodes.forEach((node) => {
        jsNodes.add(node.parentPath)
      })
    }
    // else {
    //   throw nodes[0].buildCodeFrameError(`Unknown macro ${tagName}`)
    // }
  })

  const stripNonEssentialProps =
    process.env.NODE_ENV == "production" && !opts.extract

  const jsNodesArray = Array.from(jsNodes)

  jsNodesArray.filter(isRootPath(jsNodesArray)).forEach((path) => {
    const macro = new MacroJs(babel, { i18nImportName, stripNonEssentialProps })
    if (macro.replacePath(path)) needsI18nImport = true
  })

  if (needsI18nImport) {
    addImport(
      babel.types,
      state.file.path.node,
      i18nImportModule,
      i18nImportName
    )
  }
}

;[...jsMacroTags].forEach((name) => {
  Object.defineProperty(module.exports, name, {
    get() {
      throwNonMacroContextError("@lingui/core/macro")
    },
  })
})

export default createMacro(macro, {
  configName: "lingui",
})

export * from "./types"
