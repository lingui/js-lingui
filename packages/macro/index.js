module.exports = require("@lingui/babel-plugin-lingui-macro/macro")
