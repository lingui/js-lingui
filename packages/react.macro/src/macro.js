import { createMacro } from "babel-plugin-macros"
import { implementationPlugin as TransformReactPlugin } from "@lingui/babel-plugin-transform-react"

const importsToCarryOver = ["DateFormat", "NumberFormat"]

function macro({ references, state, babel }) {
  const { types: t } = babel
  const plugin = new TransformReactPlugin(babel)

  const toKeepImports = ["Trans"]

  // Trick the plugin into thinking we've processed an import
  plugin.setImportDeclarations(
    Object.keys(references).reduce((obj, key) => {
      if (key !== "default" || key !== "Trans") obj[key] = key
      return obj
    }, {})
  )

  for (let [tagName, tags] of Object.entries(references)) {
    if (importsToCarryOver.includes(tagName)) toKeepImports.push(tagName)

    tags.forEach(openingTag => {
      if (!t.isJSXOpeningElement(openingTag.container)) return // Exclude closing elements

      const node = openingTag.context.parentPath.container

      plugin.JSXElement({ node }, state.file)
    })
  }

  const linguiReactImport = state.file.path.node.body.find(
    importNode =>
      t.isImportDeclaration(importNode) &&
      importNode.source.value === "@lingui/react"
  )

  // Handle adding the import or altering the existing import
  if (linguiReactImport) {
    toKeepImports.forEach(name => {
      if (
        linguiReactImport.specifiers.findIndex(
          specifier => specifier.imported && specifier.imported.name === name
        ) === -1
      ) {
        linguiReactImport.specifiers.push(
          t.importSpecifier(t.identifier(name), t.identifier(name))
        )
      }
    })
  } else {
    state.file.path.node.body.unshift(
      t.importDeclaration(
        toKeepImports.map(name =>
          t.importSpecifier(t.identifier(name), t.identifier(name))
        ),
        t.stringLiteral("@lingui/react")
      )
    )
  }
}

export default createMacro(macro)
// For intellisense as these will be ignored by babel-macro
// Might want to copy PropTypes here?
export const Trans = () => {}
export const Plural = () => {}
export const Select = () => {}
export const SelectOrdinal = () => {}
export const DateFormat = () => {}
export const NumberFormat = () => {}
