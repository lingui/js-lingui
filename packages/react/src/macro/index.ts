import { createMacro, MacroParams } from "babel-plugin-macros"
import { getConfig as loadConfig, LinguiConfigNormalized } from "@lingui/conf"

import { MacroJSX } from "./macroJsx"
import type { NodePath } from "@babel/traverse"
import {
  addImport,
  isRootPath,
  LinguiMacroOpts,
  throwNonMacroContextError,
} from "@lingui/macro-lib"

export const jsxMacroTags = new Set([
  "Trans",
  "Plural",
  "Select",
  "SelectOrdinal",
])

let config: LinguiConfigNormalized

// todo, make memoization on the conf library level
function getConfig(_config?: LinguiConfigNormalized) {
  if (_config) {
    config = _config
  }
  if (!config) {
    config = loadConfig()
  }
  return config
}

export function macro({ references, state, babel, config }: MacroParams) {
  const opts: LinguiMacroOpts = config as LinguiMacroOpts

  const { TransImportModule, TransImportName } = getConfig(
    opts.linguiConfig
  ).runtimeConfigModule

  const jsxNodes = new Set<NodePath>()

  Object.keys(references).forEach((tagName) => {
    const nodes = references[tagName]

    if (jsxMacroTags.has(tagName)) {
      // babel-plugin-macros return JSXIdentifier nodes.
      // Which is for every JSX element would be presented twice (opening / close)
      // Here we're taking JSXElement and dedupe it.
      nodes.forEach((node) => {
        // identifier.openingElement.jsxElement
        jsxNodes.add(node.parentPath.parentPath)
      })
    }
    // else {
    //   throw nodes[0].buildCodeFrameError(`Unknown macro ${tagName}`)
    // }
  })

  const stripNonEssentialProps =
    process.env.NODE_ENV == "production" && !opts.extract

  const jsxNodesArray = Array.from(jsxNodes)

  jsxNodesArray.filter(isRootPath(jsxNodesArray)).forEach((path) => {
    const macro = new MacroJSX(babel, { stripNonEssentialProps })
    macro.replacePath(path)
  })

  if (jsxNodes.size) {
    addImport(
      babel.types,
      state.file.path.node,
      TransImportModule,
      TransImportName
    )
  }
}

;[...jsxMacroTags].forEach((name) => {
  Object.defineProperty(module.exports, name, {
    get() {
      throwNonMacroContextError("@lingui/react/macro")
    },
  })
})

export default createMacro(macro, {
  configName: "lingui",
})

export * from "./types"
