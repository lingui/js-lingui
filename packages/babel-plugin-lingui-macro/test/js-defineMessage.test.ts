import { macroTester } from "./macroTester"
describe.skip("", () => {})

macroTester({
  cases: [
    {
      name: "defineMessage should support template literal",
      code: `
        import { defineMessage } from '@lingui/core/macro';
        const message = defineMessage\`Message\`
    `,
    },
    {
      name: "defineMessage can be called by alias `msg`",
      code: `
        import { msg } from '@lingui/core/macro';
        const message1 = msg\`Message\`
        const message2 = msg({message: "Message"})
    `,
    },
    {
      name: "should expand macros in message property",
      code: `
        import { defineMessage, plural, arg } from '@lingui/core/macro';
        const message = defineMessage({
          comment: "Description",
          message: plural(value, { one: "book", other: "books" })
        })
    `,
    },
    {
      name: "defineMessage macro could be renamed",
      code: `
        import { defineMessage as defineMessage2, plural as plural2 } from '@lingui/core/macro';
        const message = defineMessage2({
          comment: "Description",
          message: plural2(value, { one: "book", other: "books" })
        })
    `,
    },
    {
      name: "should left string message intact",
      code: `
        import { defineMessage } from '@lingui/core/macro';
        const message = defineMessage({
          message: "Message"
        })
    `,
    },
    {
      name: "should transform template literals",
      code: `
        import { defineMessage } from '@lingui/core/macro';
        const message = defineMessage({
          message: \`Message \${name}\`
        })
    `,
    },
    {
      name: "should preserve custom id",
      code: `
        import { defineMessage } from '@lingui/core/macro';
        const message = defineMessage({
          id: "msg.id",
          message: "Message"
        })
    `,
    },
    {
      name: "Production - only essential props are kept, without id",
      production: true,
      code: `
        import { defineMessage } from '@lingui/core/macro';
        const msg = defineMessage({
            message: \`Hello \${name}\`,
            comment: 'description for translators',
            context: 'My Context',
        })
    `,
    },
    {
      name: "Production - only essential props are kept",
      production: true,
      code: `
        import { defineMessage } from '@lingui/core/macro';
        const msg = defineMessage({
            message: \`Hello \${name}\`,
            id: 'msgId',
            comment: 'description for translators',
            context: 'My Context',
        })
    `,
    },
    {
      name: "should preserve values",
      code: `
        import { defineMessage, t } from '@lingui/core/macro';
        const message = defineMessage({
          message: t\`Hello \${name}\`
        })
    `,
    },
  ],
})
