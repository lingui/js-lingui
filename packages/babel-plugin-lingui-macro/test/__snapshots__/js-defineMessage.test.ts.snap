// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Production - only essential props are kept 1`] = `
import { defineMessage } from "@lingui/core/macro";
const msg = defineMessage({
  message: \`Hello \${name}\`,
  id: "msgId",
  comment: "description for translators",
  context: "My Context",
});

↓ ↓ ↓ ↓ ↓ ↓

const msg =
  /*i18n*/
  {
    id: "msgId",
    values: {
      name: name,
    },
  };

`;

exports[`Production - only essential props are kept, without id 1`] = `
import { defineMessage } from "@lingui/core/macro";
const msg = defineMessage({
  message: \`Hello \${name}\`,
  comment: "description for translators",
  context: "My Context",
});

↓ ↓ ↓ ↓ ↓ ↓

const msg =
  /*i18n*/
  {
    id: "oT92lS",
    values: {
      name: name,
    },
  };

`;

exports[`defineMessage can be called by alias \`msg\` 1`] = `
import { msg } from "@lingui/core/macro";
const message1 = msg\`Message\`;
const message2 = msg({ message: "Message" });

↓ ↓ ↓ ↓ ↓ ↓

const message1 =
  /*i18n*/
  {
    id: "xDAtGP",
    message: "Message",
  };
const message2 =
  /*i18n*/
  {
    id: "xDAtGP",
    message: "Message",
  };

`;

exports[`defineMessage macro could be renamed 1`] = `
import {
  defineMessage as defineMessage2,
  plural as plural2,
} from "@lingui/core/macro";
const message = defineMessage2({
  comment: "Description",
  message: plural2(value, { one: "book", other: "books" }),
});

↓ ↓ ↓ ↓ ↓ ↓

const message =
  /*i18n*/
  {
    id: "SlmyxX",
    message: "{value, plural, one {book} other {books}}",
    comment: "Description",
    values: {
      value: value,
    },
  };

`;

exports[`defineMessage should support template literal 1`] = `
import { defineMessage } from "@lingui/core/macro";
const message = defineMessage\`Message\`;

↓ ↓ ↓ ↓ ↓ ↓

const message =
  /*i18n*/
  {
    id: "xDAtGP",
    message: "Message",
  };

`;

exports[`should expand macros in message property 1`] = `
import { defineMessage, plural, arg } from "@lingui/core/macro";
const message = defineMessage({
  comment: "Description",
  message: plural(value, { one: "book", other: "books" }),
});

↓ ↓ ↓ ↓ ↓ ↓

const message =
  /*i18n*/
  {
    id: "SlmyxX",
    message: "{value, plural, one {book} other {books}}",
    comment: "Description",
    values: {
      value: value,
    },
  };

`;

exports[`should left string message intact 1`] = `
import { defineMessage } from "@lingui/core/macro";
const message = defineMessage({
  message: "Message",
});

↓ ↓ ↓ ↓ ↓ ↓

const message =
  /*i18n*/
  {
    id: "xDAtGP",
    message: "Message",
  };

`;

exports[`should preserve custom id 1`] = `
import { defineMessage } from "@lingui/core/macro";
const message = defineMessage({
  id: "msg.id",
  message: "Message",
});

↓ ↓ ↓ ↓ ↓ ↓

const message =
  /*i18n*/
  {
    id: "msg.id",
    message: "Message",
  };

`;

exports[`should preserve values 1`] = `
import { defineMessage, t } from "@lingui/core/macro";
const message = defineMessage({
  message: t\`Hello \${name}\`,
});

↓ ↓ ↓ ↓ ↓ ↓

const message =
  /*i18n*/
  {
    id: "OVaF9k",
    message: "Hello {name}",
    values: {
      name: name,
    },
  };

`;

exports[`should transform template literals 1`] = `
import { defineMessage } from "@lingui/core/macro";
const message = defineMessage({
  message: \`Message \${name}\`,
});

↓ ↓ ↓ ↓ ↓ ↓

const message =
  /*i18n*/
  {
    id: "A2aVLF",
    message: "Message {name}",
    values: {
      name: name,
    },
  };

`;
